go.mod: upgrade grpc-gateway to v2

tbg opened the pull request:
This PR is an offshoot from attempting to upgrade gRPC.
It turned out in the end that we can upgrade gRPC without bumping grpc-gateway to v2. However, upgrading grpc-gateway seems like a good vessel for ripping of the FUD we have around mixing gogoproto and google.golang.org/protobuf. We should do it to avoid running into thorny dep issues during routine upgrades.

tbg committed:
go.mod: upgrade gRPC to v.1.68.0

from v1.56.3 to v1.68.0.

TODO: changelog

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134964)
<!-- Reviewable:end -->

tbg committed:
remove unused annotations.proto imports

tbg committed:
BUILD.bazel: different package path for annotations pkg

Similar but different error:

```
link: package conflict error: google.golang.org/genproto/googleapis/cloud/location: package imports google.golang.org/genproto/googleapis/api/annotations
	  was compiled with: @org_golang_google_genproto//googleapis/api/annotations:annotations
	but was linked with: @org_golang_google_genproto_googleapis_api//annotations:annotations
```

Compare with the original:

```
link: package conflict error: cloud.google.com/go/pubsub/apiv1/pubsubpb: package imports google.golang.org/genproto/googleapis/api/annotations
	  was compiled with: @org_golang_google_genproto_googleapis_api//annotations:annotations
	but was linked with: @org_golang_google_genproto//googleapis/api/annotations:annotations
```

tbg committed:
also replace annotation refs in patch files

Error unchanged from last commit. I did remember to `./dev gen bazel`.

```
link: package conflict error: google.golang.org/genproto/googleapis/cloud/location: package imports google.golang.org/genproto/googleapis/api/annotations
          was compiled with: @org_golang_google_genproto//googleapis/api/annotations:annotations
        but was linked with: @org_golang_google_genproto_googleapis_api//annotations:annotations
```

tbg committed:
go.mod: go get google.golang.org/genproto@v0.0.0-20230526161137-0005af68ea54

See https://github.com/googleapis/go-genproto/issues/1015.

Sadly grpc-gateway is incompatible with this version of `genproto`:

```
ERROR: no such package '@@org_golang_google_genproto//googleapis/api/httpbody': BUILD file not found in directory 'googleapis/api/httpbody' of external repository @@org_golang_google_genproto. Add a BUILD file to a directory to mark it as a package.
ERROR: /private/var/tmp/_bazel_tbg/b1346cddcc70d57afdaa90f7f09f9b2c/external/com_github_grpc_ecosystem_grpc_gateway/runtime/BUILD.bazel:5:11: no such package '@@org_golang_google_genproto//googleapis/api/httpbody': BUILD file not found in directory 'googleapis/api/httpbody' of external repository @@org_golang_google_genproto. Add a BUILD file to a directory to mark it as a package. and referenced by '@@com_github_grpc_ecosystem_grpc_gateway//runtime:go_default_library'
```

Since we are on the final `grpc-gateway/v1` version already[^1], we'll
have to make the leap to v2 to fix this.

[^1]: https://github.com/grpc-ecosystem/grpc-gateway/releases/tag/v1.16.0

tbg committed:
*: use v2 import path for grpc-gateway

This won't compile but it's fine, just gets this change out of the way.

tbg committed:
go.mod: go get github.com/grpc-ecosystem/grpc-gateway/v2@v2.0.0

```
ERROR: /private/var/tmp/_bazel_tbg/b1346cddcc70d57afdaa90f7f09f9b2c/external/com_github_grpc_ecosystem_grpc_gateway_v2/runtime/BUILD.bazel: no such target '@@com_github_grpc_ecosystem_grpc_gateway_v2//runtime:runtime': target 'runtime' not declared in package 'runtime' defined by /private/var/tmp/_bazel_tbg/b1346cddcc70d57afdaa90f7f09f9b2c/external/com_github_grpc_ecosystem_grpc_gateway_v2/runtime/BUILD.bazel
ERROR: /Users/tbg/go/src/github.com/cockroachdb/cockroach/pkg/util/protoutil/BUILD.bazel:5:11: no such target '@@com_github_grpc_ecosystem_grpc_gateway_v2//runtime:runtime': target 'runtime' not declared in package 'runtime' defined by /private/var/tmp/_bazel_tbg/b1346cddcc70d57afdaa90f7f09f9b2c/external/com_github_grpc_ecosystem_grpc_gateway_v2/runtime/BUILD.bazel and referenced by '//pkg/util/protoutil:protoutil'
```

tbg committed:
go mod tidy

tbg committed:
./dev gen bazel --mirror (2x)

tbg committed:
BUILD.bazel: update gazelle hints for grpc-gateway to v2

Now we get:

```
ERROR: /private/var/tmp/_bazel_tbg/b1346cddcc70d57afdaa90f7f09f9b2c/external/go_googleapis/google/api/BUILD.bazel: no such target '@@go_googleapis//google/api:httpbody_go_proto': target 'httpbody_go_proto' not declared in package 'google/api' defined by /private/var/tmp/_bazel_tbg/b1346cddcc70d57afdaa90f7f09f9b2c/external/go_googleapis/google/api/BUILD.bazel (did you mean httpbody_proto, or httpbody.proto?)
ERROR: /private/var/tmp/_bazel_tbg/b1346cddcc70d57afdaa90f7f09f9b2c/external/com_github_grpc_ecosystem_grpc_gateway_v2/runtime/BUILD.bazel:5:11: no such target '@@go_googleapis//google/api:httpbody_go_proto': target 'httpbody_go_proto' not declared in package 'google/api' defined by /private/var/tmp/_bazel_tbg/b1346cddcc70d57afdaa90f7f09f9b2c/external/go_googleapis/google/api/BUILD.bazel (did you mean httpbody_proto, or httpbody.proto?) and referenced by '@@com_github_grpc_ecosystem_grpc_gateway_v2//runtime:go_default_library'
```

tbg committed:
go.mod: go get github.com/grpc-ecosystem/grpc-gateway/v2@v2.23.0

tbg committed:
go mod tidy

tbg committed:
go.mod: go get google.golang.org/genproto/googleapis/api@latest

tbg committed:
fix grpc-gatway bazel patch

tbg committed:
protoutil: update marshaler signature

tbg committed:
update BUILD.bazel files to use grpc-gateway v2 for codegen

Error after this:

```
/external/com_github_grpc_ecosystem_grpc_gateway_v2/internal/descriptor/apiconfig/BUILD.bazel:7:14:
no such package '@@googleapis//google/api':
The repository '@@googleapis' could not be resolved: Repository '@@googleapis' is not defined and referenced by '@@com_github_grpc_ecosystem_grpc_gateway_v2//internal/descriptor/apiconfig:apiconfig_proto'
```

I added a .patch trying to fix this (see this commit) but it didn't seem
to take effect.

rickystewart committed:
[WIP] continue upgrading `grpc-gateway`

Epic: none
Release note: None

tbg commented:
https://cockroachlabs.slack.com/archives/CJ0H8Q97C/p1733177280312459?thread_ts=1731420942.261169&cid=CJ0H8Q97C is the latest on how to get something that at least fails at compile time.
